### ISSUE TITLE
What's the litmus test for choosing `foreach` over `for`?

### ISSUE THREAD

**Issue by e-kayrakli** [chapel-lang/chapel#18075]
`foreach` specifies order independence in a loop. At least as of today, that's the only semantic difference between a `foreach` and a `for`.

One potentially confusing case came up in the module code here https://github.com/chapel-lang/chapel/pull/18070/files#diff-15ddf427382ddf230cfe5db25e95805d22768d605c6dee8f4efb50d5ae6ceb1cL67

There are two ways to think here:

1. The loop is over a range with a negative stride, this is a pretty specific ordering and as such we shouldn't use `foreach`
2. The loop body doesn't have any order-dependence, so the order in which the iterator yields doesn't matter. We should use `foreach`.

---

Here's a related thought experiment: You have an array-like class. Now you want to add a `reverseIterator` to this class, does it look like:

1.

```chapel
iter reverseIterator() {
  for i in myRange by -1 {
    yield myElements(i);
  }
}
```

or

2.

```chapel
iter reverseIterator() {
  foreach i in myRange by -1 {
    yield myElements(i);
  }
}
```

I'm personally on camp 2 because as the user we should give as much information as possible to the compiler. And this loop body clearly doesn't have any order-dependence. The user of the `reverseIterator` should make their own decisions about whether their use is order-dependent and whether they should use it with `for` vs `foreach`. 

---

Note that things are not too different when `for` vs `forall` are considered. In other words `forall i in 1..10` and `forall i in 1..10 by -1` are pretty much same things when it comes to ordering.

**Comment by mppf** [comment]
> The loop is over a range with a negative stride, this is a pretty specific ordering and as such we shouldn't use foreach

I guess I'm so far in the other camp that this viewpoint doesn't even make sense to me...

Why does a reverse ordering matter? If I write `foreach i in 1..10` then we are specifying a forward ordering (that being the normal way to enumerate 1..10). But we don't mind saying this can be order independent.

The idea of `foreach` is that it is a statement about that loop itself. We could take the position that iterating over a range like `myRange by -1` is not order independent (but I don't understand why we would do that - the range is still describing a set of indices we can iterate over in parallel). Anyway, if we did that, `foreach i in myRange by -1` would indicate that *this* foreach loop is happy to run in any order - but the combination with iteration over `myRange by -1` (and that not being order independent, in this hypothetical case) would mean that in practice that loop would not be order independent. (But it's still right to put `foreach` on it, because that loop is saying it is order independent - it's just if it's iterating over something not order independent then the combination after iterator inline is not order independent).

**Comment by mppf** [comment]
Now, as a style issue, if somebody writes

```
forall x in myObject.reverseIterator()
```

it would be reasonable to say "why? just use the regular iterator" and maybe that's what this issue is getting at. I'd view this kind of thing as the domain of style guidance and warnings though, rather than something to impact what foreach means.

**Comment by bradcray** [comment]
@e-kayrakli : Is the question you're asking "Should the user be able to write the `foreach` version of the iterator?" (i.e., should it be legal to do so) or "Should the compiler prevent the user from writing that version?" or "Does it matter whether or not they wrote that version?"  And when you say you're in camp 2, which do you mean?

> forall x in myObject.reverseIterator()
>
> it would be reasonable to say "why? just use the regular iterator"

I'm not sure I agree with that.  First, the parallel iterator named `reverseIterator()` could use a completely different heuristic in terms of how it creates tasks or divides up the iterations between the tasks (but I'm guessing that you're imagining that it's exactly the same as the default iterator in those respects?).  Second, because it may matter (to someone, somewhere) that the serial loops that each task executes are running backwards, even if they can't know how many tasks will be used, where the boundaries between those tasks will be, etc.  (though depending on the iterator's logic, it may be that they will be able to?)

**Comment by mppf** [comment]
> Second, because it may matter (to someone, somewhere) that the serial loops that each task executes are running backwards

Right, but it can't matter to the body of the `forall x in myObject.reverseIterator()` loop. It would have to be something in one of the iterators that was using this property. Otherwise, we would not be able to write `forall` here and would need to write `forwave` (or whatever the name of that ends up being).

**Comment by e-kayrakli** [comment]
> @e-kayrakli : Is the question you're asking "Should the user be able to write the foreach version of the iterator?" (i.e., should it be legal to do so) or "Should the compiler prevent the user from writing that version?" or "Does it matter whether or not they wrote that version?" And when you say you're in camp 2, which do you mean?

My question is mostly for the users. How should they choose between `for` and `foreach` in the scenarios I've posted above. I don't think the compiler should do anything with this regard (definitely not an error, *mayybe* we can do something with `--performance-pitfalls` here to encourage them to use `foreach` in some cases. I believe we do give out warnings with some verbose flags to say that "we aren't vectorizing this loop, because it uses `for`").

When I say I'm in camp 2. I mean both number 2s in my OP (which are saying the same thing):

> The loop body doesn't have any order-dependence, so the order in which the iterator yields doesn't matter. We should use foreach.

i.e. When I'm writing an iterator, I should choose between `for` and `foreach` only by looking at the loop body, and not trying to think about how this iterator might be used. And when using an iterator, I don't need to worry and choose between `for` and `foreach` based on my loop body.


> Second, because it may matter (to someone, somewhere) that the serial loops that each task executes are running backwards, even if they can't know how many tasks will be used, where the boundaries between those tasks will be, etc. (though depending on the iterator's logic, it may be that they will be able to?)

Note that the iterator names are intentionally muddying the waters here. Would you feel the same way if the iterator was named `forwardIterator` (akin to default `range` iterators in behavior) ? So, if you'd think the same way, then it would mean we shouldn't vectorize `forall i in 1..10`.

**Comment by bradcray** [comment]
@e-kayrakli : When you say "loop body" here:

> When I'm writing an iterator, I should choose between for and foreach only by looking at the loop body, and not trying to think about how this iterator might be used.

I think you mean "the aspects of the specific loop's body that are visible within the iterator (discounting the yield)" as opposed to the loop body for the iterator itself back at the iterator's callsite(s), right?

Thanks for clarifying the question.  I think that, with the clarifications, I'm in camp 2 as well.  I was mostly balking at what I was interpreting in Michael's response as saying "if you're using a forall, why would you ever invoke any iterator other than the default?"

**Comment by e-kayrakli** [comment]
> I think you mean "the aspects of the specific loop's body that are visible within the iterator (discounting the yield)" as opposed to the loop body for the iterator itself back at the iterator's callsite(s), right?

Yep, that's what I mean.

**Comment by mppf** [comment]
> Thanks for clarifying the question. I think that, with the clarifications, I'm in camp 2 as well. 

Sounds like the 3 of us, at least, agree on Camp 2 i.e. the loop in `iter reverseIterator` should use `foreach`.

> I was mostly balking at what I was interpreting in Michael's response as saying "if you're using a forall, why would you ever invoke any iterator other than the default?"

Sounds like this element of the discussion does not impact the decision here but we can talk more about that if you like. (I do agree that there are cases where one wants a `forall` with a non-default iterator - but how weird/rare/confusing those are is not so clear to me).
